src/hardware/pwm.rs

```diff
@@ -0,0 +1,194 @@
+    i * ((10.0 * R_SENSE) / V_PWM)
+}
+
+/// Convert maximum voltage to relative pulsewidth for the (analog voltage)
+/// max output voltage of the TEC driver.
+pub fn v_to_pwm(v: f32) -> f32 {
+    v / (4.0 * V_PWM)
+}
+
+/// TEC limit types
+///
+/// LimV - Upper and lower voltage limit
+/// LimIUp - Upper current limit
+/// LimILow - Lower current limit
+pub enum Limit {
+    LimV,
+    LimIUp,
+    LimILow,
+}
+
+/// Pwm pins.
+///
+/// lim_v<n>_pin - voltage limit pin
+/// lim_i_up<n>_pin - upper current limit pin
+/// lim_i_low<n>_pin - lower current limit pin
+/// * <n> specifies Thermostat output channel
+pub struct PwmPins {
+    pub lim_v0_pin: PE9<Alternate<AF1>>,
+    pub lim_v1_pin: PE11<Alternate<AF1>>,
+    pub lim_v2_pin: PE13<Alternate<AF1>>,
+    pub lim_v3_pin: PE14<Alternate<AF1>>,
+    pub lim_i_up0_pin: PD12<Alternate<AF2>>,
+    pub lim_i_up1_pin: PD13<Alternate<AF2>>,
+    pub lim_i_up2_pin: PD14<Alternate<AF2>>,
+    pub lim_i_up3_pin: PD15<Alternate<AF2>>,
+    pub lim_i_low0_pin: PC6<Alternate<AF2>>,
+    pub lim_i_low1_pin: PB5<Alternate<AF2>>,
+    pub lim_i_low2_pin: PC8<Alternate<AF2>>,
+    pub lim_i_low3_pin: PC9<Alternate<AF2>>,
+}
+
+type Pt0<T, S> = super::hal::pwm::Pwm<T, S, ComplementaryDisabled, ActiveHigh, ActiveHigh>;
+type Pt1<T, S> = super::hal::pwm::Pwm<T, S, ComplementaryImpossible, ActiveHigh, ActiveHigh>;
+
+/// PWM driver struct containing the PWM output pins.
+pub struct Pwm {
+    lim_v0: Pt0<TIM1, C1>,
+    lim_v1: Pt0<TIM1, C2>,
+    lim_v2: Pt0<TIM1, C3>,
+    lim_v3: Pt1<TIM1, C4>,
+    lim_i_up0: Pt1<TIM4, C1>,
+    lim_i_up1: Pt1<TIM4, C2>,
+    lim_i_up2: Pt1<TIM4, C3>,
+    lim_i_up3: Pt1<TIM4, C4>,
+    lim_i_low0: Pt1<TIM3, C1>,
+    lim_i_low1: Pt1<TIM3, C2>,
+    lim_i_low2: Pt1<TIM3, C3>,
+    lim_i_low3: Pt1<TIM3, C4>,
+}
+
+impl Pwm {
+    /// Construct a new PWM driver for all Thermostat output channels.
+    ///
+    /// # Args
+    /// * `clocks` - Reference to CoreClocks
+    /// * `tim_rec` - Peripheral Reset and Enable Control for the timer 1/2/4
+    /// * `tim` - Timer 1/2/4 peripherals
+    /// * `pins` - PWM output pins
+    pub fn new(
+        clocks: &CoreClocks,
+        tim_rec: (rec::Tim1, rec::Tim3, rec::Tim4),
+        tim: (TIM1, TIM3, TIM4),
+        pins: PwmPins,
+    ) -> Pwm {
+        fn init_pwm_pin<P: PwmPin<Duty = u16>>(pin: &mut P) {
+            pin.set_duty(0);
+            pin.enable();
+        }
+
+        let (mut lim_v0, mut lim_v1, mut lim_v2, mut lim_v3) = tim.0.pwm(
+            (
+                pins.lim_v0_pin,
+                pins.lim_v1_pin,
+                pins.lim_v2_pin,
+                pins.lim_v3_pin,
+            ),
+            F_PWM,
+            tim_rec.0,
+            clocks,
+        );
+        init_pwm_pin(&mut lim_v0);
+        init_pwm_pin(&mut lim_v1);
+        init_pwm_pin(&mut lim_v2);
+        init_pwm_pin(&mut lim_v3);
+
+        let (mut lim_i_up0, mut lim_i_up1, mut lim_i_up2, mut lim_i_up3) = tim.2.pwm(
```

Comment:
I'd first make all PWMs, then `init_pwm_pin` on all of them. Makes it more readable.